### PR TITLE
Improve MongoDB query sanitization

### PR DESCRIPTION
**Pull request by luismiramirez**
This commit changes how the MongoDB queries from the Mongo Ruby Driver
are sanitized.

Previously, the sanitizer filtered almost the whole query, not allowing
users to know which attributes and embedded documents were involved.

Also, the arrays were collapsed, hiding all the elements except the
first one. This might be problematic in Mongo queries as documents don't
have a closed schema so the sanitization could hide some attributes.

This change makes the Mongo query sanitization more similar to what we
do with SQL queries using sql_lexer.

Before:

```js
{
  "$db": "?",
  "documents": "[?]",
  "insert": "posts",
  "lsid": "?",
  "ordered": true
}
```

After:

```js
{
  "$db": "?",
  "documents": [
    {
      "_id": "?",
      "authors": [
        {
          "_id": "?",
          "name": "?"
        },
        {
          "_id": "?",
          "name": "?",
          "surname": "?"
        }
      ],
      "body": "?",
      "title": "?"
    }
  ],
  "insert": "posts",
  "lsid": "?",
  "ordered": true
}
```

Fixes: #769

### CODE REVIEW

File: .changesets/improve-mongo-ruby-driver-sanitization.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+bump: "patch"
+type: "change"
+---
+
+The MongoDB query sanitization now shows all the attributes in the query at all levels.
+Only the actual values are filtered with a `?` character.
```
Comment (tombruijn) [inline .changesets/improve-mongo-ruby-driver-sanitization.md:7]:
```suggestion
Only the actual values are filtered with a `?` character. Less MongoDB queries are now marked as N+1 queries when they weren't the exact same query.
```

File: lib/appsignal/utils/query_params_sanitizer.rb [hunk 2]
```diff
@@ -39,7 +39,7 @@ def sanitize_array(array, only_top_level, key_sanitizer)
           else
             array.map do |value|
               sanitize(value, only_top_level, key_sanitizer)
-            end.uniq
+            end
```
Comment (tombruijn) [inline lib/appsignal/utils/query_params_sanitizer.rb:42]: I'd highlight this change in the commit message, that this previously hid how many documents were inserted/updated/etc if they all had the same attributes.
It's a small change that's easily overlooked.

It's also part of the query sanitizer, which is also used by the ElasticSearch integration. I don't think it's a problem, because removing duplicate entries may hide how complex the query really was. "Did one object get inserted? Or 100.000 objects with the same structure?"

Can you add a test case for this as well? Two duplicate sanitized array entries should both appear in the sanitized result.

**Review by tombruijn — CHANGES_REQUESTED** [review]

File: lib/appsignal/utils/query_params_sanitizer.rb (hunk 2 above)
Comment (matsimitsu) [reply lib/appsignal/utils/query_params_sanitizer.rb:42]: This breaks some things down the line, though. Our processor/agent now assumes a query is the same if a hash of the body matches. 

Adding dynamic data (such as the number of documents inserted) breaks this convention and will lead to a huge increase in stored "unique" events in the database, where each permutation of the resulting array results in a new "unique" event being stored. 

Ideally, the body should be the same regardless of whether one or 1000 documents were inserted/updated, or we have to update the hashing feature to take this into account and uniq it there to ensure the resulting hash is the same for 1/2/1000 documents inserted/updated/removed.

**Review by matsimitsu — CHANGES_REQUESTED** [review]
See nested comment: https://github.com/appsignal/appsignal-ruby/pull/819/files#r800801365

File: lib/appsignal/utils/query_params_sanitizer.rb (hunk 2 above)
Comment (matsimitsu) [reply lib/appsignal/utils/query_params_sanitizer.rb:42]: > @matsimitsu the query is not the same if a different number of documents are inserted.

Ahh, ok for Mongo/Elastic it was the same, regardless of the number of documents inserted. All of this digesting was because we couldn't get the line number of the query in a performant way, back when we implemented this. 

The idea was that the event (in this case the query) is the same, regardless of arguments, e.g. the performance is about the query on line 12 of your Model, not about a combination of the query on line 12 of the model, and the arguments (1/2/3/10 documents). 

The reason is that for an app with a wildly varying number of documents inserted, we generate tons of events, this makes it more difficult for a customer to track the performance of an event, since it's now 10 events, depending on documents inserted. This also means the event now differs per sample, so in order to find the event with 10 documents inserted, you'd need to browse a ton of samples to find the right event, instead of having one event for one query at the same place in the event tree. 

This is a departure from the current implementation and expectations and may need a bit more documentation than just a changelog line.

Comment (luismiramirez) [reply lib/appsignal/utils/query_params_sanitizer.rb]: Readded the `uniq` to reduce the number of events.

File: .changesets/improve-mongo-ruby-driver-sanitization.md (hunk 1 above)
Comment (tombruijn) [reply .changesets/improve-mongo-ruby-driver-sanitization.md:7]: I've updated the changeset entry suggestion in my previous comment.

**Review by tombruijn — APPROVED** [review]

**Comment by tombruijn** [comment]
After this is merged, let's release an alpha, then install it in real mongodb app and see what the results are.

**Review by matsimitsu — APPROVED** [review]

**Comment by tombruijn** [comment]
Test is running, moving to waiting.

**Review by jeffkreeftmeijer — APPROVED** [review]
